### DISCUSSION TITLE
Trying to analyze Amstrad CPC dumps...

### DISCUSSION THREAD

**Discussion by realmml** [keirf/disk-utilities#167] [Q&A]
Hi! :)

Because disk-analyse is working great with the Amiga disk images and because there is support for IBM MFM formats, I thought that it would be a nice idea to try to support Amstrad CPC disk dumps.
The Amstrad CPC floppy disk controller is a µPD765, and even if the physical floppy is an exotic 3" format, the disk format is MFM and simple side, because the drive has only one head; meaning that you need to flip a floppy for reading the other side. And the standard not protected format is 40 tracks, 9 sectors per track and each sector is 512 bytes in size.

Then I am going to test with a typical protected game, in this case, the European release of After Burner by Activision. I have uploaded the raw dump that I am using for testing [here](https://www.dropbox.com/s/xwm9snd0jfn96w2/After%20Burner%20%28CPC%29.zip?dl=0).

This floppy has the following structure:
```
Track 0 has 8x512-byte sectors and 1 weak sector (the penultimate)
Tracks 1-2 are the usual 9x512-byte sectors 
Tracks 3-17 are 5x1024 bytes sectors
Track 18 is unformatted
Tracks 19-37 are the usual 9x512-byte sectors
Tracks 38-41 are unformatted
```

Then my first step in trying to analyse the disc is by choosing as format ibm_mfm_dd:
```
$ ./disk-analyse.exe "cpc_tests/After Burner/track" ab.ipf -v -s 0 -e 41 -S0 -f ibm_mfm_ddFound format "ibm_mfm_dd"
PLL Parameters: period_adj=5% phase_adj=60%
T0.0: IBM-MFM DD (8 512-byte sectors, 4096 bytes)
T1.0-2.0: IBM-MFM DD (9 512-byte sectors, 4608 bytes)
T3.0-17.0: IBM-MFM DD (5 1024-byte sectors, 5120 bytes)
T18.0: Unformatted
T19.0-37.0: IBM-MFM DD (9 512-byte sectors, 4608 bytes)
T38.0-41.0: Unformatted
```
And as you can see, it has done a great job, disk-analyse only has not detected the weak sector in track 0. And yes, I know that disk-analyse doesn't support create IPFs for other systems and that it is a little more complex than set the platform to 0x04 for CPC in the IPF header, but it is perfect for my tests.

Then of course, I have created a format file for defining cpc protections, for passing to disk-analyse as parameter -c. My first try is something like this:
```
default = amstrad_cpc

amstrad_cpc
    * ibm_mfm_dd 

after_burner
    0 ibm_mfm_dd # This is the one with the weak sector (I don't know how to express that)
    1-2 ibm_mfm_dd 
    3-17 ibm_mfm_dd 
    18 ignore
    19-37 ibm_mfm_dd 
    * ignore
```

Then when I use:
```
$ ./disk-analyse.exe "cpc_tests/After Burner/track" ab.ipf -v -s 0 -e 41 -S0 -c cpc_protections.txt -f amstrad_cpc
```
Logically get the same information about the disk format.

But if I try to choose my after_burner floppy definition (and yes, I know that it is wrong and surely I will need to add extra code in libdisk/format/ibm.c and libdisk/format/ibm_img.c), y results are a lot different:
```
$ ./disk-analyse.exe "cpc_tests/After Burner/track" ab.ipf -v -s 0 -e 41 -S0 -c cpc_protections.txt -f after_burnerFound format "after_burner"
PLL Parameters: period_adj=5% phase_adj=60%
T0.0: IBM-MFM DD (8 512-byte sectors, 4096 bytes)
T1.0-2.0: IBM-MFM DD (9 512-byte sectors, 4608 bytes)
T3.0-8.0: IBM-MFM DD (5 1024-byte sectors, 5120 bytes)
T9.0: Unformatted
T10.0-17.0: IBM-MFM DD (5 1024-byte sectors, 5120 bytes)
T18.0-41.0: Unformatted
```
It is detecting track 9 and tracks 19-37 as unformatted.

Then, I would like to get some directions about which should be my next steps for continuing my research about this subject of supporting Amstrad CPC floppy protections. 

Thanks!
Mauri.

PS: And of course, I have tested unprotected floppies and disk-analyse works great:
```
 ./disk-analyse.exe "cpc_tests/Airwolf/track" airwolf.ipf -v -s 0 -e 41 -S0 -f ibm_mfm_dd
Found format "ibm_mfm_dd"
PLL Parameters: period_adj=5% phase_adj=60%
T0.0-39.0: IBM-MFM DD (9 512-byte sectors, 4608 bytes)
T40.0-41.0: Unformatted
```

**Comment by keirf** [comment]
Have you ever tried samdisk? It's the best for Sam, Spectrum and CPC. And these days should be able to analyse HFE image files. So if you can convert to HFE this is a great way to go.

**Comment by keirf** [comment]
I say this because copy protected IBM tracks will need new or extended analysers in disk-utilities. So you can get some of the way but not all the way.

**Comment by realmml** [comment]
Of course, Keir, we use samdisk for preservation since always, hehehe. But there are limitations, for example, it doesn't support HFE v3 files, for example. 

Another tool that we used too is SugarConvDsk, but this program is not as well supported or updated as Disk-Analyse (last update was precovid).

Aside from that, the standard disk format used for preservation in the CPC, is the EDSK disk image format. And this format is really bad for this task, it was developed in the second half of the 90s and a few extensions have been added to it since then for supporting special protections, but it is a hack over a hack. It doesn't support fluxes at all.

It doesn't help that the number of CPC emulators supporting something different than EDSK is practically zero. There are only two cpc emulators supporting IPF, and only one of them is updated frequently. Making it really hard to test those IPFs.

Then because that and because the number of protections in the CPC world is much more limited than in Amiga (the base for all the cpc protections are weak sectors and/or using sectors and tracks with not standard size) I think it is a good idea to bring the CPC world to disk-analyze (or to the present of preservation, hehehe). 

I feel that all the work that you have done about Amiga protections, it can be used as a base to support those CPC protections. And yes, this is going to need new analysers for handling copy protected tracks in the IBM MFM format, I would like to work in that task, even if support IBM MFM protected tracks in disk-analyse is thing for the future, but because that I want to ask for your opinion about the subject.

I would need some direction about this task; and being sincere, the CPC would need a new format for preservation, based in fluxes and with a section having the protections used in the tracks. I would think that IPF would be a nice one, but maybe with your experience, could you recommend a better/simpler format for preserving IBM MFM tracks with protection?

Thanks,
Mauri.
